storage.c

```diff
@@ -15,7 +16,453 @@
+    }
+    if (new_it == NULL)
+        return -1;
+    // so we can free the chunk on a miss
+    new_it->slabs_clsid = clsid;
+
+    io_pending_storage_t *p = do_cache_alloc(c->thread->io_cache);
+    // this is a re-cast structure, so assert that we never outsize it.
+    assert(sizeof(io_pending_t) >= sizeof(io_pending_storage_t));
+    memset(p, 0, sizeof(io_pending_storage_t));
+    p->active = true;
+    p->miss = false;
+    p->badcrc = false;
+    p->noreply = c->noreply;
+    // io_pending owns the reference for this object now.
+    p->hdr_it = it;
+    p->resp = resp;
+    p->q = q; // quicker access to the queue structure.
+    obj_io *eio = &p->io_ctx;
+
+    // FIXME: error handling.
+    if (chunked) {
+        unsigned int ciovcnt = 0;
+        size_t remain = new_it->nbytes;
+        item_chunk *chunk = (item_chunk *) ITEM_schunk(new_it);
+        // TODO: This might make sense as a _global_ cache vs a per-thread.
+        // but we still can't load objects requiring > IOV_MAX iovs.
+        // In the meantime, these objects are rare/slow enough that
+        // malloc/freeing a statically sized object won't cause us much pain.
+        eio->iov = malloc(sizeof(struct iovec) * IOV_MAX);
+        if (eio->iov == NULL) {
+            item_remove(new_it);
+            do_cache_free(c->thread->io_cache, p);
+            return -1;
+        }
+
+        // fill the header so we can get the full data + crc back.
+        eio->iov[0].iov_base = new_it;
+        eio->iov[0].iov_len = ITEM_ntotal(new_it) - new_it->nbytes;
+        ciovcnt++;
+
+        while (remain > 0) {
+            chunk = do_item_alloc_chunk(chunk, remain);
+            // FIXME: _pure evil_, silently erroring if item is too large.
+            if (chunk == NULL || ciovcnt > IOV_MAX-1) {
+                item_remove(new_it);
+                free(eio->iov);
+                // TODO: wrapper function for freeing up an io wrap?
+                eio->iov = NULL;
+                do_cache_free(c->thread->io_cache, p);
+                return -1;
+            }
+            eio->iov[ciovcnt].iov_base = chunk->data;
+            eio->iov[ciovcnt].iov_len = (remain < chunk->size) ? remain : chunk->size;
+            chunk->used = (remain < chunk->size) ? remain : chunk->size;
+            remain -= chunk->size;
+            ciovcnt++;
+        }
+
+        eio->iovcnt = ciovcnt;
+    }
+
+    // Chunked or non chunked we reserve a response iov here.
+    p->iovec_data = resp->iovcnt;
+    int iovtotal = (c->protocol == binary_prot) ? it->nbytes - 2 : it->nbytes;
+    if (chunked) {
+        resp_add_chunked_iov(resp, new_it, iovtotal);
+    } else {
+        resp_add_iov(resp, "", iovtotal);
+    }
+
+    // We can't bail out anymore, so mc_resp owns the IO from here.
+    resp->io_pending = (io_pending_t *)p;
+
+    eio->buf = (void *)new_it;
+    p->c = c;
+
+    // We need to stack the sub-struct IO's together for submission.
+    eio->next = q->stack_ctx;
```

Comment:
can you help me understand these two lines 333-334?
Can we ever have q->stack_ctx having a value other than NULL?

Reply:
Here is my understanding -
for a given connection, if multiple read requests arrive at the same time before we could complete the request, we keep on adding the previous q->stack_ctx to eio->next and then update the q->stack_ctx to eio.

The extstore subsystem will handle all the IO's pointed to by _obj_io and the stack_ctx always points to the head of this list. Is this correct? 

How are we protected in the event that this _obj_io is being handled by the extstore subsystem and _obj_io is not meddled with?

Reply:
not sure I understand the question but I'll detail the ownership process:
- a network read fills an input buffer
- commands are read out of the input buffer. requests are stacked into `q->stack_ctx`
- if end of buffer or max reqs per yied hit, switch to send result data
- if queue has stack, submit stack to extstore
- client connection processing suspends
- extstore works through all submitted obj_io's from that connection. Once the queue hits 0, the connection is signaled to resume
- worker thread resumes client connection, processes the result callbacks for each obj_io that was submitted.
- etc

So the ownership of the _whole_ queue passes between worker thread and extstore IO threads. There shouldn't be any time when both are able to access them.

Reply:
I think I found the answer to my question - client requests will not be accepted on that connection until we call redispatch_conn. So we are guaranteed not to have _obj_io getting modified while being processed by the extstore_io_thread.

Reply:
Yes. That's what I said in my detail here too. This happens either during an ascii multiget (get key key key), or any time a client submits multiple requests at once (can be unrelated gets/metagets). Rest of the time it doesn't/can't batch.

There's a complication that I added to make proxy work, in that code running on a deferred queue can add new deferred IO's to its stack. IE: in proxy's case a request needs to be able to queue up a retry while it's technically running in a callback already. That's why there are two levels of callbacks (pre and post cleanup). Might be a bit hard to follow.

Reply:
Yes, the details you posted above make it much clearer to follow. (I guess you beat me to it in responding sooner :) ). I've a much better understanding of the deferred IO queue subsystem and the extstore code paths now.  Thank you!

I will have start reviewing the proxy PR now  - https://github.com/memcached/memcached/pull/716. 
I am sure I will understand the complication you mentioned above once I start reading the code.

Reply:
ok, cool! Keep in mind proxy is under a lot of work at the moment :) I also frequently rebase/squash/etc the branch.